cliquet/tests/resource/test_events.py

```diff
@@ -0,0 +1,87 @@
+from contextlib import contextmanager
+
+import webtest
+from pyramid.config import Configurator
+
+from cliquet.events import ResourceChanged, ACTIONS
+from cliquet.tests.testapp import main as testapp
+from cliquet.tests.support import unittest, BaseWebTest, get_request_class
+
+
+@contextmanager
+def notif_broken(app):
+    old = app.registry.notify
+
+    def buggy(event):
+        if not isinstance(event, ResourceChanged):
+            return old(event)
+        raise Exception("boom")
+
+    app.registry.notify = buggy
+    yield
+    app.registry.notify = old
+
+
+class EventsTest(BaseWebTest, unittest.TestCase):
+    def setUp(self):
+        super(EventsTest, self).setUp()
+        self.events = []
+        self.config.add_subscriber(self.listener, ResourceChanged)
+        self.config.commit()
+        self.body = {'data': {'name': 'de Paris'}}
+
+    def tearDown(self):
+        self.events = []
+        super(EventsTest, self).tearDown()
+
+    def listener(self, event):
+        self.events.append(event)
+
+    def get_test_app(self, settings=None):
+        settings = self.get_app_settings(settings)
+        self.config = Configurator(settings=settings)
+        app = testapp(config=self.config)
+        app = webtest.TestApp(app)
+        app.RequestClass = get_request_class(self.api_prefix)
+        return app
+
+    def test_event_triggered_on_write(self):
+        self.app.post_json(self.collection_url, self.body,
+                           headers=self.headers, status=201)
+        self.assertEqual(len(self.events), 1)
+        self.assertEqual(self.events[0].payload['action'], ACTIONS.CREATE)
+
+    def test_event_triggered_on_update(self):
+        resp = self.app.post_json(self.collection_url, self.body,
+                                  headers=self.headers, status=201)
+        record = resp.json['data']
+        record_url = self.get_item_url(record['id'])
+
+        self.app.patch_json(record_url, self.body, headers=self.headers,
+                            status=200)
+        self.assertEqual(len(self.events), 2)
+        self.assertEqual(self.events[0].payload['action'], ACTIONS.CREATE)
+        self.assertEqual(self.events[1].payload['action'], ACTIONS.UPDATE)
+
+    def test_event_triggered_on_delete(self):
+        resp = self.app.post_json(self.collection_url, self.body,
+                                  headers=self.headers, status=201)
+        record = resp.json['data']
+        record_url = self.get_item_url(record['id'])
+
+        self.app.delete(record_url, headers=self.headers, status=200)
+        self.assertEqual(len(self.events), 2)
+        self.assertEqual(self.events[0].payload['action'], ACTIONS.CREATE)
+        self.assertEqual(self.events[1].payload['action'], ACTIONS.DELETE)
+
```

Comment:
Three missing test cases IHMO:
* No event when POST on collection that returns 200 (existing id provided)
* PUT that create
* PUT that replace

:)